wasmlib/helpers/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+        Err(err) => Err(err),
+    }
+}
+
+/// Given a pointer to the start of a byte array of WAC data
+/// encode the data into the codec
+///
+///
+/// # Safety
+///
+/// This function assumes that the load_raw_block function returns valid results
+pub unsafe fn load_wac_block_caller(blk_cid: Cid) -> Result<Box<[u8]>, libipld::error::Error> {
+    let blk_cid_bytes = blk_cid.to_bytes();
+    let cidptr = blk_cid_bytes.as_ptr();
+
+    let blk_res = &*load_wac_block(cidptr, blk_cid_bytes.len() as u8);
+    let res = blk_res.to_result();
+    match res {
+        Ok(v) => Ok(v.to_owned().into_boxed_slice()),
+        Err(err) => Err(err),
+    }
+}
+
+use std::ffi::c_void;
+
+use libipld::Cid;
+
+pub trait Codec {
+    /// Given a pointer to the start of a byte array of WAC data
+    /// encode the data into the codec
+    ///
+    /// # Safety
+    ///
+    /// This function assumes the block pointer has size have been allocated and filled.
+    unsafe fn encode(&self, ptr: *mut u8, len: u32) -> *const ValueOrError;
+
+    /// Given a pointer to the start of a byte array and
+    /// its length, decode it into a standard IPLD codec representation
+    /// (for now WAC)
+    ///
+    /// # Safety
+    ///
+    /// This function assumes the block pointer has size have been allocated and filled.
+    unsafe fn decode(&self, ptr: *mut u8, len: u32) -> *const ValueOrError;
+}
+
+pub trait AdlBase {
+    /// Takes a pointer and length of a byte array containing WAC encoded data and returns
+    /// a pointer to the ADL instance.
+    ///
+    /// # Safety
+    ///
+    /// This function assumes the block pointer has size have been allocated and filled.
+    unsafe fn load_adl(&self, ptr: *mut u8, len: u32) -> *mut ADLorWAC;
+}
+
+#[allow(clippy::missing_safety_doc)]
+pub trait AdlBytes {
+    unsafe fn new_bytes_reader(&self, adlptr: *const c_void) -> *mut c_void;
+    unsafe fn read_adl(&self, adlptr: *mut u8, bufptr: *mut u8, bufleni: i32) -> *const ReadResp;
+    unsafe fn seek_adl(&self, adlptr: *mut c_void, offset: i64, whence: u32) -> u64;
+}
+
+#[allow(clippy::missing_safety_doc)]
+pub trait AdlMap {
+    unsafe fn new_map_iter(&self, adlptr: *const c_void) -> *mut c_void;
+    unsafe fn iter_next(&self, adlptr: *mut c_void) -> *const IterResp;
+    unsafe fn get_key(
+        &self,
+        adlptr: *mut c_void,
+        key_ptr: *mut u8,
+        key_len: i32,
+    ) -> *const ADLorWAC;
+    unsafe fn adl_len(&self, adlptr: *const c_void) -> i64;
+}
+
+// unsafe fn iter_done(adlptr: *mut u8) -> *const BoolOrError;
+
+#[allow(clippy::missing_safety_doc)]
+pub trait AdlList {
+    unsafe fn get_index(&self, adlptr: *mut u8, index: i32) -> *const ADLorWAC;
+    unsafe fn new_list_iter(&self, adlptr: *mut u8) -> *const ValueOrError;
+    unsafe fn iter_next(&self, adlptr: *mut c_void) -> *const IterResp;
+    unsafe fn adl_len(&self, adlptr: *const c_void) -> i64;
+}
+
+/*
+/// Given a pointer to the start of a byte array of WAC data
```

Comment:
This is mostly me being bad at Rust/WASM things.

What I want:
1. A way to specify the ABI
2. A way to make it easy for people to make new codecs/ADLs and just fill in the blanks, like fulfilling an interface contract
3. I want my code to compile

The `#[no_mangle]` public functions are part of the ABI, but I can't figure out how I'd include them in a new module since I don't know how to get the new module to make data available to the public function. Suggestions welcome, but I'd prefer not to have to make a WASM "init" call to set things up and IIUC that doesn't happen automatically like it might in some other environments.

So this is basically the code you can copy-paste to fulfill the ABI in your new WASM module. You fulfill the traits and then have the `#[no_mangle]` functions call out to your trait implementation.

Better ideas super duper welcome